peps/pep-9999.rst

```diff
@@ -0,0 +1,632 @@
+a safe version which performs error checking (for example,
+``PyTuple_GET_ITEM`` vs ``PyTuple_GetItem``). It may help to
+be able to group them into their own tiers - the "unsafe API" tier and
+the "safe API" tier
+[`Issue 61 <https://github.com/capi-workgroup/problems/issues/61>`__].
+
+Use of the C Language
+---------------------
+
+A number of issues were raised with respect to the way that CPython
+uses the C language. First there is the issue of which C dialect
+we use, and how we test our compatibility with it
+[`Issue 42 <https://github.com/capi-workgroup/problems/issues/42>`__].
+
+Usage of ``const`` in the API is currently sparse, but it is not
+clear whether this is something that we should consider changing
+[`Issue 38 <https://github.com/capi-workgroup/problems/issues/38>`__].
+
+We currently use the C types ``long`` and ``int``, where ``stdint``
+and ``int32_t`` would have been better choices
+[`Issue 27 <https://github.com/capi-workgroup/problems/issues/27>`__].
+
+We are using C language features which are hard for other languages
+to interact with
+[`Issue 35 <https://github.com/capi-workgroup/problems/issues/35>`__].
+
+There are API functions that take a ``PyObject*`` arg which must be
+of a more specific type (such as ``PyTuple_Size``, which fails if
+its arg is not a ``PyTupleObject*``). It is an open question whether this
+is a good pattern to have, or whether the API should expect the
+more specific type
+[`Issue 31 <https://github.com/capi-workgroup/problems/issues/31>`__].
+
+There are functions in the API that take concrete types, such as
+``PyDict_GetItemString`` which performs a dictionary lookup for a key
+specified as a c string rather than ``PyObject*``. At the same time,
+for ``PyDict_ContainsString`` it is not considered appropriate to
+add a concrete type alternative. The principle around this should
+be documented in the guidelines
+[`Issue 23 <https://github.com/capi-workgroup/problems/issues/23>`__].
+
+Implementation Flaws
+--------------------
+
+Below is a list of localized implementation flaws. Most of these can
+probably be fixed incrementally, if we choose to do so. They should,
+in any case, be avoided in any new API design.
+
+There are functions that don't follow the convention of
+returning ``0`` for success and ``-1`` for failure. For
+example, ``PyArg_ParseTuple`` returns 0 for success and
+non-zero for failure
+[`Issue 25 <https://github.com/capi-workgroup/problems/issues/25>`__].
+
+The macros ``Py_CLEAR`` and ``Py_SETREF`` access their arg more than
+once, so if the arg is an expression with side effects, they are
+duplicated
+[`Issue 3 <https://github.com/capi-workgroup/problems/issues/3>`__].
+
+The meaning of ``Py_SIZE`` depends on the type and is not always
+reliable
+[`Issue 10 <https://github.com/capi-workgroup/problems/issues/10>`__].
+
+**Naming**
+
+``PyLong`` and ``PyUnicode`` use names which don't match the python
+types they represent (int/str). This can be fixed in a new API
```

Comment:
```suggestion
``PyLong`` and ``PyUnicode`` use names which don't match the Python
types they represent (``int``/``str``). This can be fixed in a new API
```